Deprovision tenant hosts affected by cloud event

olaaun opened the pull request.

olaaun committed:
Deprovision tenant host affected by cloud event

andreer reviewed (approved):
LGTM, but I suggest getting an additional reviewer on this to be on the safe side 👍

freva commented on controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/CloudEventReporter.java:
```diff
@@ -27,33 +32,67 @@
+    }
+
+    private void submitIssue(CloudEvent event) {
+        if (event.affectedInstances.isEmpty())
```
If this is possible, this needs to be moved to the parent method so we can skip `deprovisionHosts()` as well.

olaaun replied:
The idea was to remove instances from affectedInstances if they were deprovisioned. I can make it more clear, by letting `deprovisionHosts` return the list of deprovisioned hosts, and skip ticket submission if `event.affectedInstances.size() == deprovisionedHosts.size()`

freva replied:
Sounds good. Event should be immutable.

freva commented on controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/CloudEventReporter.java:
```diff
@@ -27,33 +32,67 @@
+                    .map(node -> node.hostname().value())
+                    .forEach(hostname -> {
+                        log.info(String.format("Setting host %s to wantToRetire and wantToDeprovision", hostname));
+                        nodeRepository.retireAndDeprovision(zone.getId(), hostname);
```
Since you are iterating over multiple zones, I assume this will throw if you try to do this on the wrong zone?

olaaun replied:
If this is done in the wrong zone, all hosts should've been filtered away by `shouldDeprovisionHost`

freva commented on controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/CloudEventReporter.java:
```diff
@@ -27,33 +32,67 @@
     }
 
+    private void deprovisionHosts(String awsRegion, CloudEvent event) {
+        for (var zone : zonesByCloudNativeRegion.get(awsRegion)) {
```
Instance name contains zone name in it, so this is not really needed.

freva commented on controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/CloudEventReporter.java:
```diff
@@ -27,33 +32,67 @@
+        }
+    }
+
+    private Predicate<Node> shouldDeprovisionHost(CloudEvent event) {
```
Also check if the node already has `wantToDeprovision` and `wantToRetire` set already.

freva commented on controller-server/src/test/java/com/yahoo/vespa/hosted/controller/maintenance/CloudEventReporterTest.java:
```diff
@@ -0,0 +1,161 @@
+
+    }
+
+    private void equalListsIgnoreOrder(List<?> l1, List<?> l2) {
```
Consider returning `Set` from `getHostnames()` and use `assertEquals(Set.of(..), getHostnames(...));`

olaaun committed:
Add wantToDeprovision and wantToRetire fields to Node

olaaun committed:
Check if number of affected instances is equal to number of deprovisioned hosts.
Check if host already wants to deprovision/retire
Change from list to set in test
